org.hl7.fhir.publisher.core/src/main/java/org/hl7/fhir/igtools/publisher/AIProcessor.java

```diff
@@ -0,0 +1,429 @@
+import org.eclipse.persistence.internal.sessions.DirectCollectionChangeRecord;
+import org.hl7.fhir.exceptions.FHIRException;
+import org.hl7.fhir.igtools.publisher.utils.OldIGAIGenerator;
+import org.hl7.fhir.r5.model.ImplementationGuide;
+import org.hl7.fhir.utilities.*;
+import org.hl7.fhir.utilities.filesystem.ManagedFileAccess;
+import org.hl7.fhir.utilities.json.model.JsonElement;
+import org.hl7.fhir.utilities.json.model.JsonObject;
+import org.hl7.fhir.utilities.json.parser.JsonParser;
+import org.hl7.fhir.utilities.npm.NpmPackage;
+import org.hl7.fhir.utilities.xhtml.NodeType;
+import org.hl7.fhir.utilities.xhtml.XhtmlNode;
+import org.hl7.fhir.utilities.xhtml.XhtmlParser;
+import org.hl7.fhir.utilities.xhtml.XhtmlToMarkdownConverter;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.*;
+
+public class AIProcessor {
+  private final String rootDir;
+
+  public AIProcessor(String rootDir) {
+    this.rootDir = rootDir;
+  }
+
+  public void processNewTemplates(List<String> langs) throws IOException {
+    // processing new templates is different to old template because of the layout with regard to
+    // languages, and because there's a specific .ai file
+    NpmPackage npm = NpmPackage.fromPackage(new FileInputStream(Utilities.path(rootDir, "package.tgz")));
+
+    for (String lang : langs) {
+      processFolderNew(new File(Utilities.path(rootDir, lang)), npm, lang);
+    }
+  }
+
+  public void processOldTemplates() throws IOException {
+    NpmPackage npm = NpmPackage.fromPackage(new FileInputStream(Utilities.path(rootDir, "package.tgz")));
+    processFolderOld(new File(rootDir), npm);
+  }
+
+  public void processFolderNew(File dir, NpmPackage npm, String lang) throws IOException {
+    List<LoadedResource> resources = new ArrayList<>();
+    List<Page> pages = new ArrayList<>();
+
+    // list all the resources in the IG, and note their type and whether they're an example
+    loadResources(dir, resources, npm, true);
+
+    // now list all the pages in the IG, and decide whether they are a page, a resource, or an alternate view of a resource
+    long oldSize = loadPages(dir, pages, resources);
+    long newSize = 0;
+
+    ZipGenerator zip = new ZipGenerator(Utilities.path(dir, "ai.zip"));
+    StringBuilder llms = new StringBuilder();
+    llms.append("# " + npm.vid() + ": " + npm.title() + "("+lang+")\r\n\r\n");
+    llms.append("## Pages\r\n\r\n");
+    for (Page p : pages) {
+      if (p.f.getName().equals("index.html")) {
+        newSize += produceMDForPage(llms, p, zip, npm);
+      }
+    }
+    for (Page p : pages) {
+      if (!p.f.getName().equals("index.html") && p.kind == PageKind.PAGE) {
+        newSize += produceMDForPage(llms, p, zip, npm);
+      }
+    }
+    llms.append("\r\n## Resources\r\n");
+    resources.sort(new LoadedResourceSorter());
+    String t = null;
+    for (LoadedResource r : resources) {
+      Page p = getPageForResource(r, pages);
```

Comment:
I get a warning saying p could be null at this point. Should there be an assert or check here?